spec/abstract/InvocationFuns_A.thy

```diff
@@ -0,0 +1,379 @@
+(*
+ * Copyright 2014, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(*
+Monad instantiations, handling of faults, errors, and interrupts.
+*)
+
+chapter "Basic Kernel and Exception Monads"
+
+theory InvocationFuns_A
+imports Ipc_A
+begin
+
+section \<open>Allow preemption at this point.\<close>
+definition
+  preemption_point :: "(unit,'z::state_ext) p_monad" where
+ "preemption_point \<equiv> doE liftE $ do_extended_op update_work_units;
+                         OR_choiceE (work_units_limit_reached)
+                           (doE liftE $ do_extended_op reset_work_units;
+                                irq_opt \<leftarrow> liftE $ do_machine_op (getActiveIRQ True);
+                                case_option (do
+                                               update_time_stamp;
+                                               result \<leftarrow> check_budget;
+                                               if (\<not> result)
+                                               then returnOk ()
+                                               else (throwError $ ())
```

Comment:
Shouldn't the "then" and "else" branches be swapped?

Reply:
Oops, yes, I think you are right. Actually, I meant to add the "cur_sc is active" check as well.

Reply:
Oh, I thought you were wanting me to do that when you said that I would be carrying on this work. Happy for you to do that, of course!